3.7.2+ Union Type Merged to Single Type in Wrapper

**TypeScript Version:**
3.8.0-dev.20191119
3.7.2

**Search Terms:**
Redux
Connect
Union

**Code**

```ts
const mapStateToProps = (state?: {}) => !state ? {} : { foo: 1 };

type Props = ReturnType<typeof mapStateToProps>;

const MyComponent = (props: Props) => null;

// Simplified from react
type FunctionComponent<P> = (props: P & { children: {} }) => null;

// Simplified from react-redux
type Matching<InjectedProps, DecorationTargetProps> = {
    [P in keyof DecorationTargetProps]: P extends keyof InjectedProps
        ? InjectedProps[P] extends DecorationTargetProps[P]
            ? DecorationTargetProps[P]
            : InjectedProps[P]
        : DecorationTargetProps[P];
};

// Simplified from react-redux
type InferableComponentEnhancerWithProps =
    <C extends FunctionComponent<Matching<Props, Props>>>(
        component: C
    ) => never;

(window as any as InferableComponentEnhancerWithProps)(MyComponent);
```

**Expected behavior:**
Code should compile without error.  This works on 3.6.3 and below.

**Actual behavior:**
```
demo.ts:25:56 - error TS2345: Argument of type '(props: { foo?: undefined; } | { foo: number; }) => null' is not assignable to parameter of type 'FunctionComponent<Matching<{ foo?: undefined; } | { foo: number; }, { foo?: undefined; }> | Matching<{ foo?: undefined; } | { foo: number; }, { foo: number; }>>'.
  Types of parameters 'props' and 'props' are incompatible.
    Type '(Matching<{ foo?: undefined; } | { foo: number; }, { foo?: undefined; }> & { children: {}; }) | (Matching<{ foo?: undefined; } | { foo: number; }, { foo: number; }> & { children: {}; })' is not assignable to type '{ foo?: undefined; } | { foo: number; }'.
      Type 'Matching<{ foo?: undefined; } | { foo: number; }, { foo?: undefined; }> & { children: {}; }' is not assignable to type '{ foo?: undefined; } | { foo: number; }'.
        Type 'Matching<{ foo?: undefined; } | { foo: number; }, { foo?: undefined; }> & { children: {}; }' is not assignable to type '{ foo: number; }'.
          Types of property 'foo' are incompatible.
            Type 'number | undefined' is not assignable to type 'number'.
              Type 'undefined' is not assignable to type 'number'.
```

We encountered this in a very large React/Redux project after attempting to update to 3.7.2.
I simplified the repro case as best I could.
Apologies for the title, I don't know the best terminology for this situation.


**Playground Link:**
[Link](https://www.typescriptlang.org/play/index.html#code/MYewdgzgLgBAtgQwA4GUoKgUwCogAoBOISEMAvDABTQaYD8AXDAN4C+AlOQHwwCENWGHRasYTZjABmIEEwCMMVgG4AUCqgBPJJhiFipCgCVMUAK4Ew2LZgA8m7SEnxkaWrj0kuqlaEiwAshoAwiBwSOCYYLAUlEhEJEweEJxkPGCmADYZ3gD0OTAoAJZhGYWShZgAJlJEcDAEmAjAUOrWMABipmDNheAhYRFRNng8MXH6iTAAZCwwwAAWhRmVDWDiohzcMOlZuflFJWUV1ZK19Y3NALQNlaYAHq3aMP4YC4VgAOY2AJJgAFaYZpVJIAGhgABFASACBhepYEAQPiYkqMWCoYBiYABtPAwd4wADWmA0jghUJhUDh2ARSKgSQAupNMHcsGBKqQiSSnL8AUDKkl0ZihcIeYCsPz4hAcfSYMzWeyyaAKVSacjJdLBUKtUJFdDYeBqYi1foNdrtUxRXykqazUxIUr9fCjXT1Xh6aplGo8gVikhSuUqjVQucmlBrlV7o8dL9JJgYQAjDKYfrhMCRKAAUTA8wQ3TjAHVClB5klyJqYDYgrKWZEFZ1upS+qFU+mbC8oG9PsNJWCUVwuJRy0LQAM01EmEFyyk0pgAG5x7yUADu70qICXMAQpFzGk3pBjcYQieTzcGmezueABaLJcl7EogRTZ-YSiAA)
Note that the playground doesn't show a behavior change from 3.7.2 to 3.6.3, but VS Code and the command-line `tsc` do.

**Related Issues:**
#33872 - Seems similar and broke at the same time

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -16540,7 +16540,7 @@ namespace ts {
                             if (!targetProperty) continue outer;
                             if (sourceProperty === targetProperty) continue;
                             // We compare the source property to the target in the context of a single discriminant type.
-                            const related = propertyRelatedTo(source, target, sourceProperty, targetProperty, _ => combination[i], /*reportErrors*/ false, IntersectionState.None);
+                            const related = propertyRelatedTo(source, target, sourceProperty, targetProperty, _ => combination[i], /*reportErrors*/ false, IntersectionState.None, /*skipOptional*/ strictNullChecks || relation === comparableRelation);
                             // If the target property could not be found, or if the properties were not related,
                             // then this constituent is not a match.
                             if (!related) {
@@ -16638,7 +16638,7 @@ namespace ts {
                 }
             }
 
-            function propertyRelatedTo(source: Type, target: Type, sourceProp: Symbol, targetProp: Symbol, getTypeOfSourceProperty: (sym: Symbol) => Type, reportErrors: boolean, intersectionState: IntersectionState): Ternary {
+            function propertyRelatedTo(source: Type, target: Type, sourceProp: Symbol, targetProp: Symbol, getTypeOfSourceProperty: (sym: Symbol) => Type, reportErrors: boolean, intersectionState: IntersectionState, skipOptional: boolean): Ternary {
                 const sourcePropFlags = getDeclarationModifierFlagsFromSymbol(sourceProp);
                 const targetPropFlags = getDeclarationModifierFlagsFromSymbol(targetProp);
                 if (sourcePropFlags & ModifierFlags.Private || targetPropFlags & ModifierFlags.Private) {
@@ -16681,7 +16681,7 @@ namespace ts {
                     return Ternary.False;
                 }
                 // When checking for comparability, be more lenient with optional properties.
-                if (relation !== comparableRelation && sourceProp.flags & SymbolFlags.Optional && !(targetProp.flags & SymbolFlags.Optional)) {
+                if (!skipOptional && sourceProp.flags & SymbolFlags.Optional && !(targetProp.flags & SymbolFlags.Optional)) {
                     // TypeScript 1.0 spec (April 2014): 3.8.3
                     // S is a subtype of a type T, and T is a supertype of S if ...
                     // S' and T are object types and, for each member M in T..
@@ -16811,7 +16811,7 @@ namespace ts {
                     if (!(targetProp.flags & SymbolFlags.Prototype) && (!numericNamesOnly || isNumericLiteralName(name) || name === "length")) {
                         const sourceProp = getPropertyOfType(source, name);
                         if (sourceProp && sourceProp !== targetProp) {
-                            const related = propertyRelatedTo(source, target, sourceProp, targetProp, getTypeOfSymbol, reportErrors, intersectionState);
+                            const related = propertyRelatedTo(source, target, sourceProp, targetProp, getTypeOfSymbol, reportErrors, intersectionState, relation === comparableRelation);
                             if (!related) {
                                 return Ternary.False;
                             }

--- a/tests/baselines/reference/unionRelationshipCheckPasses.js
+++ b/tests/baselines/reference/unionRelationshipCheckPasses.js
@@ -1,0 +1,7 @@
+//// [unionRelationshipCheckPasses.ts]
+const item: { foo?: undefined } | { foo: number } = null as any as { foo?: number | undefined };
+
+
+//// [unionRelationshipCheckPasses.js]
+"use strict";
+var item = null;

--- a/tests/baselines/reference/unionRelationshipCheckPasses.symbols
+++ b/tests/baselines/reference/unionRelationshipCheckPasses.symbols
@@ -1,0 +1,7 @@
+=== tests/cases/compiler/unionRelationshipCheckPasses.ts ===
+const item: { foo?: undefined } | { foo: number } = null as any as { foo?: number | undefined };
+>item : Symbol(item, Decl(unionRelationshipCheckPasses.ts, 0, 5))
+>foo : Symbol(foo, Decl(unionRelationshipCheckPasses.ts, 0, 13))
+>foo : Symbol(foo, Decl(unionRelationshipCheckPasses.ts, 0, 35))
+>foo : Symbol(foo, Decl(unionRelationshipCheckPasses.ts, 0, 68))
+

--- a/tests/baselines/reference/unionRelationshipCheckPasses.types
+++ b/tests/baselines/reference/unionRelationshipCheckPasses.types
@@ -1,0 +1,10 @@
+=== tests/cases/compiler/unionRelationshipCheckPasses.ts ===
+const item: { foo?: undefined } | { foo: number } = null as any as { foo?: number | undefined };
+>item : { foo?: undefined; } | { foo: number; }
+>foo : undefined
+>foo : number
+>null as any as { foo?: number | undefined } : { foo?: number | undefined; }
+>null as any : any
+>null : null
+>foo : number | undefined
+

--- a/tests/cases/compiler/unionRelationshipCheckPasses.ts
+++ b/tests/cases/compiler/unionRelationshipCheckPasses.ts
@@ -1,0 +1,2 @@
+// @strict: true
+const item: { foo?: undefined } | { foo: number } = null as any as { foo?: number | undefined };

